File: apps/api/utils/email.py
from __future__ import annotations

import os


def load_disposable_domains() -> set[str]:
    file_path = os.path.abspath(
        os.path.join(
            os.path.dirname(__file__),
            "..",
            "..",
            "..",
            "..",
            "src",
            "lib",
            "data",
            "disposable-email-domains.txt",
        )
    )
    domains: set[str] = set()
    if not os.path.exists(file_path):
        return domains
    try:
        with open(file_path, encoding="utf-8") as f:
            for line in f:
                d = line.strip().lower()
                if d:
                    domains.add(d)
    except Exception:
        pass
    return domains


DISPOSABLE_DOMAINS: set[str] = load_disposable_domains()


def is_disposable_email_domain(email: str) -> bool:
    email = email.strip().lower()
    at_idx = email.rfind("@")
    if at_idx < 0 or at_idx == len(email) - 1:
        return False
    domain = email[at_idx + 1 :]
    if domain in DISPOSABLE_DOMAINS:
        return True
    parts = domain.split(".")
    return any(".".join(parts[idx:]) in DISPOSABLE_DOMAINS for idx in range(1, len(parts) - 1))

File: apps/api/src/platform/email.ts
import { readFileSync } from 'node:fs'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'

/**
 * Disposable-email-domain screening.
 *
 * Ported from `utils/email.py`. The list is loaded once at module scope from a
 * newline-delimited data file, exactly as the Python does — this is static I/O
 * whose result is identical for every request, so reading it per call would buy
 * nothing (`.claude/rules/performance-server-side.md` §3.5).
 *
 * The file is optional. `utils/email.py` resolves it to a path that does not
 * exist in this repository, so `DISPOSABLE_DOMAINS` is empty in the running
 * FastAPI service and `is_disposable_email_domain` always returns `false`
 * today. That behaviour is reproduced rather than "fixed": populating the list
 * here would start rejecting sign-ups the service currently accepts, which is a
 * product decision and not part of a port.
 */

const DATA_FILE = join(
  dirname(fileURLToPath(import.meta.url)),
  '..',
  'data',
  'disposable-email-domains.txt'
)

function loadDisposableDomains(): ReadonlySet<string> {
  const domains = new Set<string>()
  try {
    for (const line of readFileSync(DATA_FILE, 'utf-8').split('\n')) {
      const domain = line.trim().toLowerCase()
      if (domain) domains.add(domain)
    }
  } catch {
    // An absent or unreadable file means "no domains are blocked", matching the
    // Python, which swallows the same failure.
  }
  return domains
}

export const DISPOSABLE_DOMAINS = loadDisposableDomains()

/**
 * Whether an address's domain — or any of its parent domains — is disposable.
 *
 * The parent walk stops one label short of the public suffix, so a two-label
 * domain is only matched in full: `mail.example.com` tests `example.com` but
 * never the bare `com`.
 */
export function isDisposableEmailDomain(
  email: string,
  domains: ReadonlySet<string> = DISPOSABLE_DOMAINS
): boolean {
  const normalized = email.trim().toLowerCase()
  const atIndex = normalized.lastIndexOf('@')
  if (atIndex < 0 || atIndex === normalized.length - 1) return false

  const domain = normalized.slice(atIndex + 1)
  if (domains.has(domain)) return true

  const parts = domain.split('.')
  for (let index = 1; index < parts.length - 1; index += 1) {
    if (domains.has(parts.slice(index).join('.'))) return true
  }
  return false
}
